src/repr/src/relation.rs

```diff
@@ -718,19 +883,221 @@ impl RelationDescBuilder {
+        };
+
+        RelationDesc {
+            typ: relation_type,
+            metadata: column_metas,
+        }
+    }
+
+    pub fn latest_version(&self) -> RelationVersion {
+        self.inner
+            .metadata
+            .values()
+            // N.B. Dropped is always greater than added.
+            .map(|meta| meta.dropped.unwrap_or(meta.added))
+            .max()
+            // If there aren't any columns we're implicitly the root version.
+            .unwrap_or(RelationVersion::root())
+    }
+
+    /// Validates internal contraints of the [`RelationDesc`] are correct.
+    ///
+    /// # Panics
+    ///
+    /// Panics if a constraint is not satisfied.
+    fn validate(&self) {
+        fn validate_inner(desc: &RelationDesc) -> Result<(), anyhow::Error> {
+            if desc.typ.column_types.len() != desc.metadata.len() {
+                anyhow::bail!("mismatch between number of types and metadatas");
+            }
+
+            for (col_idx, meta) in &desc.metadata {
+                if col_idx.0 > desc.metadata.len() {
+                    anyhow::bail!("column index out of bounds");
+                }
+                if meta.added >= meta.dropped.unwrap_or(RelationVersion(u64::MAX)) {
+                    anyhow::bail!("column was added after it was dropped?");
+                }
+                if desc.typ().columns().get(meta.typ_idx).is_none() {
+                    anyhow::bail!("typ_idx incorrect");
+                }
+            }
+
+            let versions = desc
+                .metadata
+                .values()
+                .map(|meta| meta.dropped.unwrap_or(meta.added));
+            let mut max = 0;
+            let mut sum = 0;
+            for version in versions {
+                max = std::cmp::max(max, version.0);
+                sum += version.0;
+            }
+
+            // Other than RelationVersion(0), we should never have duplicate
+            // versions and they should always increase by 1. In other words, the
+            // sum of all RelationVersions should be the sum of [0, max].
+            //
+            // N.B. n * (n + 1) / 2 = sum of [0, n]
+            //
+            // While I normally don't like tricks like this, it allows us to
+            // validate that our column versions are correct in O(n) time and
+            // without allocations.
+            if sum != (max * (max + 1) / 2) {
+                anyhow::bail!("there is a duplicate or missing relation version");
+            }
+
+            Ok(())
+        }
+
+        assert_ok!(validate_inner(&self.inner), "validate failed! {self:?}");
+    }
+}
+
 /// Diffs that can be generated proptest and applied to a [`RelationDesc`] to
 /// exercise schema migrations.
 #[derive(Debug)]
 pub enum PropRelationDescDiff {
     AddColumn { name: ColumnName, typ: ColumnType },
     DropColumn { name: ColumnName },
-    MovePosition { name: ColumnName, new_pos: usize },
```

Comment:
This is only test code, after this change we can't really move the position of a column